lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2077,12 +2077,62 @@ bool normalizeWeights(Module *M, ConvolutionNode &CV,
   return true;
 }
 
+/// Gets Constant or returns nullptr if input is not Constant.
+/// Skips QuantizeNode if present.
+static Constant *getConstant(const NodeValue &NV) {
+  Node *N = NV.getNode();
+  if (isa<QuantizeNode>(N)) {
+    N = N->getNthInput(QuantizeNode::InputIdx);
+  }
+  return dyn_cast<Constant>(N);
+}
+
 bool OptimizeBatchNorm::run(Function *F, const CompilationContext &cctx) {
   LOG_SCOPE(F->getLogContext(), getName());
   bool changed = false;
   auto &nodes = F->getNodes();
   auto *M = F->getParent();
 
+  // For each node:
+  for (auto &node : nodes) {
+    // Remove BN if mean,var,eps,scale,beta values make it redundant as per
+    // expression (X - mean) * (1.0 / sqrt(var + eps)) * scale + bias.
+    if (auto *BN = dyn_cast<BatchNormalizationNode>(&node)) {
+      auto *scaleC = getConstant(BN->getScale());
+      auto *biasC = getConstant(BN->getBias());
+      auto *meanC = getConstant(BN->getMean());
+      auto *varC = getConstant(BN->getVar());
+      if (!(scaleC && biasC && meanC && varC)) {
+        continue;
+      }
+      float scale, bias, mean, var;
+      if ((!isUniformConstant<float>(*scaleC, scale)) ||
+          (!isUniformConstant<float>(*biasC, bias)) ||
+          (!isUniformConstant<float>(*meanC, mean)) ||
+          (!isUniformConstant<float>(*varC, var))) {
+        continue;
+      }
+      float eps = BN->getEpsilon();
+
+      // Redundancy check based on strict equalities.
+      if (scale == 1.0f && bias == 0.0f && mean == 0.0f &&
+          (var + eps == 1.0f)) {
+        BN->getResult().replaceAllUsesOfWith(BN->getInput());
+        changed = true;
+        continue;
+      }
```

Comment:
This should be covered by the relaxed redundancy check below, right? Do we need it?